### PR TITLE
tikv_kv: support tablet level snapshot acquisition

### LINKED ISSUE #13050: Support tablet level snapshot acquisition.
We should provide tablet level snapshot acquisition in multi-rocksdb version,

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #13050

What's Changed:

```
1. Refactor raftstore v1's local reader to make ReadDelegate implement ReadExecutor
2. Implement the relevant code for raftstore v2.
3. Add a test `test_snap_cache_hit` in raftstore v1.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Please add a release note.

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with None.
If this PR will be picked to release branch, then a release note is probably required.
```

### CODE REVIEW

File: components/raftstore/src/store/worker/read.rs [hunk 1]
```diff
@@ -692,6 +713,7 @@ where
             // Forward to raftstore.
             Ok(None) => self.redirect(RaftCommand::new(req, cb)),
             Err(e) => {
+                println!("{:?}", e);
```
Comment (zhangjinpeng1987) [inline components/raftstore/src/store/worker/read.rs:716]: Please remove this verbose print.

Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs:716]: I forget to remove it. Thanks for the reminding.

File: components/raftstore/src/store/worker/read.rs [hunk 2]
```diff
@@ -441,25 +449,35 @@ where
...
+        self.factory.open_tablet_cache_latest(region_id).unwrap()
     }
 
-    fn get_snapshot(&mut self, create_time: Option<ThreadReadId>) -> Arc<E::Snapshot> {
+    fn get_snapshot(
+        &mut self,
+        create_time: Option<ThreadReadId>,
+        mut region_id: u64,
+    ) -> Arc<E::Snapshot> {
+        if self.factory.get_factory_version() == TabletFactoryVersion::Single {
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/read.rs:461]: I don't see the point of this check. If it's single rocksdb, it can just ignore `region_id`.

File: components/engine_test/src/lib.rs [hunk 3]
```diff
@@ -129,13 +130,15 @@ pub mod kv {
             db_opt: Option<DBOptions>,
             cfs: &[&str],
             opts: Option<Vec<CFOptions>>,
+            multi_rocksdb: bool,
```
Comment (BusyJay) [inline components/engine_test/src/lib.rs:133]: Why the change? If it's single rocksdb, then it should use `KvEngineFactory`.

File: components/raftstore/src/store/worker/read.rs (hunk 2 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs:461]: Now, open_tablet_cache_any of KvEngineFactory does not ignore the region_id. Ok, maybe I should change it rather than check here.

File: components/engine_test/src/lib.rs (hunk 3 above)
Comment (SpadeA-Tang) [reply components/engine_test/src/lib.rs:133]: KvEngineFactory uses RocksEngine whereas TestTabletFactory uses KvTestEngine.

Comment (BusyJay) [reply components/engine_test/src/lib.rs:133]: Then you can just implement `TestTabletFactory` and `TestTabletFactoryV2` just like `KvEngineFactory` and `KvEngineFactoryV2`.

Comment (SpadeA-Tang) [reply components/engine_test/src/lib.rs:133]: For the consistency with real implementation? Okay, I will do it later.

File: components/raftstore/src/store/worker/read.rs [hunk 4]
```diff
@@ -468,14 +477,16 @@ where
...
+        store_meta: Arc<Mutex<StoreMeta>>,
+        router: C,
+    ) -> Self {
         LocalReader {
             store_meta,
-            kv_engine,
+            factory,
             router,
-            snap_cache: None,
-            cache_read_id,
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs:478]: It is binded with snap to be a tuple in snap_cache: `snap_cache: HashMap<u64, (ThreadReadId, Arc<E::Snapshot>)>`.

File: components/raftstore/src/store/worker/read.rs [hunk 5]
```diff
@@ -722,7 +733,45 @@ where
     }
 
     pub fn release_snapshot_cache(&mut self) {
-        self.snap_cache.take();
+        self.snap_cache.clear();
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs:736]: Okay, I get your point. I will ix it.

File: components/engine_test/src/lib.rs [hunk 6]
```diff
@@ -205,6 +309,30 @@ pub mod kv {
             None
         }
 
+        fn open_tablet_cache_latest(&self, id: u64) -> Option<KvTestEngine> {
```
Comment (tonyxuqqi) [inline components/engine_test/src/lib.rs:312]: The performance of this API is pretty bad----it requires O(N) to query all the keys to find the latest one. 
Maybe it's better to have a map <region_id, latest_suffix> for a quick search. 
Given its usage scenario is on the hot read path, the performance of this API is critical.

File: src/server/engine_factory_v2.rs [hunk 7]
```diff
@@ -80,6 +80,31 @@ impl TabletFactory<RocksEngine> for KvEngineFactoryV2 {
         None
     }
 
+    fn open_tablet_cache_latest(&self, id: u64) -> Option<RocksEngine> {
```
Comment (tonyxuqqi) [inline src/server/engine_factory_v2.rs:83]: The performance of this API is pretty bad----it requires O(N) to query all the keys to find the latest one.
Maybe it's better to have a map <region_id, latest_suffix> for a quick search.
Given its usage scenario is on the hot read path, the performance of this API is critical.

Comment (SpadeA-Tang) [reply src/server/engine_factory_v2.rs:83]: Good catch and good suggestion.

File: components/raftstore/src/store/worker/read.rs [hunk 8]
```diff
@@ -441,25 +449,27 @@ where
     C: ProposalRouter<E::Snapshot> + CasualRouter<E>,
     E: KvEngine,
 {
-    fn get_engine(&self) -> &E {
-        &self.kv_engine
+    fn get_tablet(&self, region_id: u64) -> E {
+        self.factory.open_tablet_cache_latest(region_id).unwrap()
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/read.rs:453]: It's not efficient to lock every time.

**Review by BusyJay — COMMENTED** [review]
How about extracting common logic as functions and create new read worker in v2? Using trait here seems too complicated.

File: components/raftstore/src/store/worker/read.rs [hunk 9]
```diff
@@ -784,25 +784,25 @@ impl<'r, 'm> RequestInspector for Inspector<'r, 'm> {
 const METRICS_FLUSH_INTERVAL: u64 = 15_000; // 15s
 
 #[derive(Clone)]
-struct ReadMetrics {
-    local_executed_requests: u64,
-    local_executed_stale_read_requests: u64,
-    local_executed_snapshot_cache_hit: u64,
+pub struct ReadMetrics {
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs:787]: Fine.

File: components/raftstore-v2/src/operation/read/read.rs [hunk 10]
```diff
@@ -0,0 +1,1302 @@
...
+/// 1. For v2's ReadDelegate (the wrapper of v1's ReadDelegate and CachedTablet): Unlike v1, which uses a single
+/// global kv rocksdb, each region in v2 has it's own kv rocksdb, namely the tablet. Equipping ReadDelegate with a
+/// CachedTablet makes tablet/snapshot acquisition very quickly. But CachedTablet requires &mut self to acquire tablet
+/// as sometimes tablet may be updated which makes CachedTablet to mutate itself. ReadDelegate is read only, so wrapping them
+/// together to workaround this.
+/// 2. For ReadDelgateWithMetric (the wrapper of ReadMetrics and v2's ReadDelegate): Unlike v1 where the LocalReader implements
+/// ReadExecutor, in v2, we use ReadDelegateWithMetric to implement ReadDelegateWithMetric so that each delegate can get its
+/// tablet/snapshot very quickly. But we also need to update some metrics which is a field of LocalReader, so we use this wrapper
+/// which will be used as a temporary local variable to make compiler happy.
+pub struct ReadDelegateWithMetric<'a, E>
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/read/read.rs:91]: Good suggestion.

File: components/raftstore-v2/src/operation/read/read.rs [hunk 11]
```diff
@@ -0,0 +1,1302 @@
...
+pub struct LocalReader<C, E>
+where
+    C: ProposalRouter<E::Snapshot> + CasualRouter<E>,
+    E: KvEngine,
+{
+    store_id: Cell<Option<u64>>,
+    store_meta: Arc<Mutex<StoreMeta<E>>>,
+    metrics: ReadMetrics,
+    // region id -> ReadDelegate
+    // The use of `Arc` here is a workaround, see the comment at `get_delegate`
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/read/read.rs:321]: Fine.

File: components/raftstore/src/store/worker/read.rs [hunk 12]
```diff
@@ -418,63 +524,71 @@ impl Progress {
     }
 }
 
-pub struct LocalReader<C, E>
+pub struct LocalReader<C, E, D, S>
```
Comment (tonyxuqqi) [inline components/raftstore/src/store/worker/read.rs:527]: please add a comment to indicate that LocalReader is the entry class to handle read all tablets's data. (Basically it's per store's reader)

File: components/raftstore-v2/src/operation/read/read.rs [hunk 13]
```diff
@@ -0,0 +1,792 @@
...
+/// updated which makes CachedTablet to mutate itself. ReadDelegate is read
+/// only, so wrapping them together to workaround this.
+///
+/// 2. For ReadDelgateWithMetric (the wrapper of ReadMetrics and v2's
+/// ReadDelegate): Unlike v1 where the LocalReader implements ReadExecutor, in
+/// v2, we use ReadDelegateExt to implement ReadDelegateExt so that each
+/// delegate can get its tablet/snapshot very quickly. But we also need to
+/// update some metrics which is a field of LocalReader, so we use this wrapper
+/// which will be used as a temporary local variable to make compiler happy.
+// pub struct ReadDelegateExt<'a, E>
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/read/read.rs:100]: I will modify it. It's the decription of the previous implementation.

File: components/raftstore/src/store/worker/read.rs (hunk 12 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs:527]: Fine.

**Review by tonyxuqqi — APPROVED** [review]

File: components/raftstore/src/store/worker/read.rs [hunk 14]
```diff
@@ -165,13 +173,126 @@ pub struct ReadDelegate {
...
+    fn clone(&self) -> Self {
+        CachedReadDelegate {
+            delegate: Arc::clone(&self.delegate),
+            kv_engine: self.kv_engine.clone(),
+        }
+    }
+}
+
+/// ReadDelegateExt combines some LocalReader's fields for temporary usage.
+pub struct ReadDelegateExt<'a, E>
```
Comment (tabokie) [inline components/raftstore/src/store/worker/read.rs:211]: How about `LocalReadContext`? `Ext` is usually reserved for trait.

File: components/raftstore/src/store/worker/read.rs [hunk 15]
```diff
@@ -786,26 +933,27 @@ impl<'r, 'm> RequestInspector for Inspector<'r, 'm> {
 
 const METRICS_FLUSH_INTERVAL: u64 = 15_000; // 15s
 
+/// The reason for it to be pub is that it is shared with raftstore v2.
```
Comment (tabokie) [inline components/raftstore/src/store/worker/read.rs:936]: I think @tonyxuqqi 's original suggestion is better, "#[RaftstoreCommon]" makes it easy to search. Ideally when the v2 works are all finished, we could gather and re-export these things in a separate namespace.

File: src/server/engine_factory_v2.rs [hunk 16]
```diff
@@ -15,8 +15,10 @@ const TOMBSTONE_MARK: &str = "TOMBSTONE_TABLET";
 
 #[derive(Clone)]
 pub struct KvEngineFactoryV2 {
-    inner: KvEngineFactory,
-    registry: Arc<Mutex<HashMap<(u64, u64), RocksEngine>>>,
+    pub inner: KvEngineFactory,
```
Comment (tabokie) [inline src/server/engine_factory_v2.rs:18]: Why `pub`?

File: src/server/engine_factory_v2.rs [hunk 17]
```diff
@@ -15,8 +15,10 @@ const TOMBSTONE_MARK: &str = "TOMBSTONE_TABLET";
...
+    pub registry: Arc<Mutex<HashMap<(u64, u64), RocksEngine>>>,
+    // registry_latest stores tablet with the latest prefix for each region
+    pub registry_latest: Arc<Mutex<HashMap<u64, (u64, RocksEngine)>>>,
```
Comment (tabokie) [inline src/server/engine_factory_v2.rs:21]: Comment the lock order constraint.

File: components/raftstore/src/store/worker/read.rs [hunk 18]
```diff
@@ -165,13 +173,126 @@ pub struct ReadDelegate {
...
+}
+
 impl Drop for ReadDelegate {
     fn drop(&mut self) {
         // call `inc` to notify the source `ReadDelegate` is dropped
         self.track_ver.inc();
     }
 }
 
+pub trait DelegateStore<E>: Send + Clone + 'static
```
Comment (tabokie) [inline components/raftstore/src/store/worker/read.rs:227]: I feel like there are too many XxDelegate types around. `ReadExecutorProvider` is probably more meaningful here.

File: components/raftstore/src/store/worker/read.rs (hunk 15 above)
Comment (tabokie) [reply components/raftstore/src/store/worker/read.rs:936]: And BTW, there seems to be a lot of shared structs/traits not annotated.

File: src/server/engine_factory.rs [hunk 19]
```diff
@@ -89,6 +91,18 @@ impl KvEngineFactoryBuilder {
             compact_event_sender: self.compact_event_sender.clone(),
         }
     }
+
+    pub fn buildv2(self) -> KvEngineFactoryV2 {
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply src/server/engine_factory.rs:95]: Good catch.

File: components/raftstore-v2/src/fsm/store.rs [hunk 20]
```diff
@@ -1,12 +1,39 @@
...
 use tikv_util::mpsc::{self, LooseBoundedSender, Receiver};
 
-use crate::{batch::StoreContext, StoreMsg};
+use crate::{batch::StoreContext, tablet::CachedTablet, StoreMsg};
+
+pub struct StoreMeta<E>
+where
+    E: KvEngine,
+{
+    /// store id
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/fsm/store.rs:17]: Good catch.

File: components/tikv_util/src/time.rs [hunk 21]
```diff
@@ -512,6 +512,13 @@ pub struct ThreadReadId {
     pub create_time: Timespec,
 }
 
+impl ThreadReadId {
+    pub fn update(&mut self, read_id: ThreadReadId) {
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/tikv_util/src/time.rs:516]: Good catch.

File: components/raftstore/src/store/worker/read.rs (hunk 14 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs:211]: Very good suggestion.

File: components/raftstore/src/store/worker/read.rs [hunk 22]
```diff
@@ -922,9 +1070,11 @@ mod tests {
     use std::{sync::mpsc::*, thread};
 
     use crossbeam::channel::TrySendError;
+    use engine_rocks::util::get_cf_handle;
     use engine_test::kv::{KvTestEngine, KvTestSnapshot};
-    use engine_traits::ALL_CFS;
+    use engine_traits::{Peekable, ALL_CFS, CF_DEFAULT};
     use kvproto::raft_cmdpb::*;
+    use rocksdb::Writable;
```
[the remark this answers is not in the document]
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs:1077]: Okay, I will modify it.

File: src/server/engine_factory_v2.rs (hunk 17 above)
Comment (SpadeA-Tang) [reply src/server/engine_factory_v2.rs:21]: Fine.

File: src/server/engine_factory_v2.rs (hunk 16 above)
Comment (SpadeA-Tang) [reply src/server/engine_factory_v2.rs:18]: build_v2 uses it to create an instance. I think maybe I should create a new for V2 and make inner private.

File: components/raftstore/Cargo.toml [hunk 23]
```diff
@@ -91,6 +91,11 @@ txn_types = { path = "../txn_types", default-features = false }
 uuid = { version = "0.8.1", features = ["serde", "v4"] }
 yatp = { git = "https://github.com/tikv/yatp.git", branch = "master" }
 
+[dependencies.rocksdb]
```
Comment (tabokie) [inline components/raftstore/Cargo.toml:94]: Remove this.

File: components/raftstore/src/router.rs [hunk 24]
```diff
@@ -168,12 +168,22 @@ where
 }
 
 /// A router that routes messages to the raftstore
-pub struct ServerRaftStoreRouter<EK: KvEngine, ER: RaftEngine> {
+#[allow(clippy::type_complexity)]
```
Comment (tabokie) [inline components/raftstore/src/router.rs:171]: No need anymore after #13190

File: components/raftstore/src/store/worker/read.rs [hunk 25]
```diff
@@ -923,8 +1076,9 @@ mod tests {
 
     use crossbeam::channel::TrySendError;
     use engine_test::kv::{KvTestEngine, KvTestSnapshot};
-    use engine_traits::ALL_CFS;
+    use engine_traits::{Peekable, SyncMutable, ALL_CFS, CF_DEFAULT};
     use kvproto::raft_cmdpb::*;
+    // use rocksdb::Writable;
```
Comment (tabokie) [inline components/raftstore/src/store/worker/read.rs:1081]: Remove this.

File: components/raftstore-v2/Cargo.toml [hunk 26]
```diff
@@ -32,18 +32,29 @@ cloud-azure = ["raftstore/cloud-azure"]
...
 raft-proto = { version = "0.7.0" }
 raftstore = { path = "../raftstore" }
 slog = "2.3"
 smallvec = "1.4"
+tikv_kv = { path = "../tikv_kv", default-features = false }
 tikv_util = { path = "../tikv_util", default-features = false }
+time = "0.1"
+txn_types = { path = "../txn_types", default-features = false }
+
+[dependencies.rocksdb]
```
Comment (tabokie) [inline components/raftstore-v2/Cargo.toml:54]: And this.

File: components/raftstore/src/store/worker/read.rs [hunk 27]
```diff
@@ -165,13 +174,128 @@ pub struct ReadDelegate {
...
+/// #[RaftstoreCommon]
+pub trait ReadExecutorProvider<E>: Send + Clone + 'static
+where
+    E: KvEngine,
+{
+    type Delegate: ReadExecutor<E>;
+
+    fn store_id(&self) -> Option<u64>;
+
+    fn get_delegate_and_len(&self, region_id: u64) -> (usize, Option<Self::Delegate>);
```
Comment (tabokie) [inline components/raftstore/src/store/worker/read.rs:237]: Add some comment to explain what is "len".

Comment (tabokie) [inline components/raftstore/src/store/worker/read.rs:237]:
```suggestion
    type Executor: ReadExecutor<E>;

    fn store_id(&self) -> Option<u64>;

    fn get_executor_and_len(&self, region_id: u64) -> (usize, Option<Self::Executor>);
```

File: components/raftstore/src/store/worker/read.rs [hunk 28]
```diff
@@ -165,13 +174,128 @@ pub struct ReadDelegate {
...
+{
+    pub fn new(store_meta: Arc<Mutex<StoreMeta>>, kv_engine: E) -> Self {
+        StoreMetaDelegate {
+            store_meta,
+            kv_engine,
+        }
+    }
+}
+
+impl<E> Deref for StoreMetaDelegate<E>
```
Comment (tabokie) [inline components/raftstore/src/store/worker/read.rs:261]: I don't feel like this is particularly necessary. It it's only used between L272-L296, let's remove it.

File: components/raftstore/Cargo.toml (hunk 23 above)
Comment (SpadeA-Tang) [reply components/raftstore/Cargo.toml:94]: Good catch.

File: components/raftstore/src/store/worker/read.rs (hunk 25 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs:1081]: Good catch.

File: components/raftstore-v2/Cargo.toml (hunk 26 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/Cargo.toml:54]: Good catch.

File: components/raftstore/src/store/worker/read.rs (hunk 27 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs:237]: Fine.

File: components/raftstore/src/store/worker/read.rs (hunk 28 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/store/worker/read.rs]: Done.

File: components/raftstore/src/router.rs (hunk 24 above)
Comment (SpadeA-Tang) [reply components/raftstore/src/router.rs]: Done.

**Review by tabokie — APPROVED** [review]

File: components/raftstore-v2/src/operation/read/read.rs [hunk 29]
```diff
@@ -0,0 +1,256 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
```
Comment (BusyJay) [inline components/raftstore-v2/src/operation/read/read.rs:1]: Why create a read file inside a read module?

File: components/raftstore-v2/Cargo.toml [hunk 30]
```diff
@@ -32,18 +32,24 @@ cloud-azure = ["raftstore/cloud-azure"]
 batch-system = { path = "../batch-system", default-features = false }
 collections = { path = "../collections" }
 crossbeam = "0.8"
+engine_rocks = { path = "../engine_rocks", default-features = false }
```
Comment (BusyJay) [inline components/raftstore-v2/Cargo.toml:35]: You should not make it as a dependency as we are supposed to generic over engine trait.

File: components/raftstore-v2/Cargo.toml [hunk 31]
```diff
@@ -32,18 +32,24 @@ cloud-azure = ["raftstore/cloud-azure"]
...
+keys = { path = "../keys", default-features = false }
 kvproto = { git = "https://github.com/pingcap/kvproto.git" }
+log_wrappers = { path = "../log_wrappers" }
 pd_client = { path = "../pd_client" }
 raft = { version = "0.7.0", default-features = false, features = ["protobuf-codec"] }
 raft-proto = { version = "0.7.0" }
 raftstore = { path = "../raftstore" }
 slog = "2.3"
 smallvec = "1.4"
+tikv_kv = { path = "../tikv_kv", default-features = false }
```
Comment (BusyJay) [inline components/raftstore-v2/Cargo.toml:49]: Why depend on `tikv_kv`? raftstore-v2 is supposed to be an implementation of `tikv_kv`.

**Comment by BusyJay** [comment]
Are these 700+ lines of code really reviewed with cautions? 🤣
